### PR TITLE
HID Switch to force sending of duplicate data and cleanup of JS API functions

### PR DESCRIPTION
**Pull request by JoergAtGithub**
I added a switch to force the sending of HID OutputReports with identical data, as needed by @Be-ing to support the S4 MK3 mapping: https://github.com/mixxxdj/mixxx/pull/4585#discussion_r813208966

After looking at the JavaScript API, I found out, that
`Q_INVOKABLE void send(const QList<int>& data, unsigned int length = 0) override`
is only there to comply with the inheritance of controller.c. But this function is practically useless. It doesn't send raw bytes as the siblings for MIDI and BULK, because this is a functionality, which the HID protocoll doesn't support. Instead it sends a OutputReport with hardcoded ReportID 0. I wonder, if this function was ever used anywhere.

Than there is
`Q_INVOKABLE void send(const QList<int>& data, unsigned int length, unsigned int reportID)`
which has already one unused argument (length).

Both send functions use QLIST as argument, while since #4521, the other functions of the JavaScript API use QByteArray.

I added a new API function with QByteArray
`Q_INVOKABLE void sendOutputReport(unsigned int reportID, const QByteArray& dataArray, resendUnchangedReport = false)`
and extended the old send function backward compatibility:
`Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length, unsigned int reportID, resendUnchangedReport = false)`
Note: I found no way to make an overloaded function work, which support both QList or QByteArray. Independend of the type in JavaScript, always the same function was called. 

Furthermore I noticed, that the comments describing these JavaScript API functions are no longer apply since #4521, I rewrote them (in Doxygen syntax). Especially the returned array of getInputReport became incompatible to the incommingdata function. And this compatibility was the only reason to prepend the reportId to the data array.

Now the HID API functions are:

**Unchanged since 2.3:**
`Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length = 0) override`
IMHO this function makes no sense, but I kept it for backward compatibility

**Extended since 2.3:**
`Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length, quint8 reportID, bool resendUnchangedReport = false)`
Added optional argument resendUnchangedReport - Note that the Length argument is unused and it uses QList

**New in 2.4 and now unified in content and type of the data array:**
`Q_INVOKABLE void sendOutputReport(quint8 reportID, const QByteArray& dataArray, bool resendUnchangedReport = false)`
`Q_INVOKABLE QByteArray getInputReport(quint8 reportID)`
`Q_INVOKABLE void sendFeatureReport(quint8 reportID, const QByteArray& reportData)`
`Q_INVOKABLE QByteArray getFeatureReport(quint8 reportID)`

This is now similar to the basic IO functions in all other HID APIs I know - as well in naming as in order of arguments

### CODE REVIEW

File: src/controllers/hid/hidcontroller.h [hunk 1]
```diff
@@ -60,21 +57,58 @@ class HidControllerJSProxy : public ControllerJSProxy {
...
 
-    Q_INVOKABLE void send(const QList<int>& data, unsigned int length, unsigned int reportID) {
-        m_pHidController->sendReport(data, length, reportID);
+    /// @brief Sends an OutputReport to HID device
+    /// @param dataArray Data to send as byte array (Javascript type Uint8Array)
+    /// @param reportID 1...255 for HID devices that uses ReportIDs - or 0 for devices, which don't use ReportIDs
+    /// @param skipIdenticalReports If set, the sending of the report is inhibited, when the data didn't changed
+    Q_INVOKABLE void sendOutputReport(const QByteArray& dataArray,
+            unsigned int reportID,
+            bool skipIdenticalReports = true) {
```
Comment (uklotzde) [inline src/controllers/hid/hidcontroller.h:94]: `resendUnchangedReport = false`

- "identity" is a different concept
- using `false` as the default value is more common
- `true` triggers an action -> positive logic
- only applies to a single report, not multiple

File: src/controllers/hid/hidiooutputreport.cpp [hunk 2]
```diff
@@ -80,7 +82,7 @@ bool HidIoOutputReport::sendCachedData(QMutex* pHidDeviceAndPollMutex,
         return false;
     }
 
-    if (!m_lastSentData.compare(m_cachedData)) {
+    if (!m_skipIdenticalReports && !m_lastSentData.compare(m_cachedData)) {
```
Comment (uklotzde) [inline src/controllers/hid/hidiooutputreport.cpp:85]: This condition is wrong

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by Be-ing** [comment]
@JoergAtGithub could you take care of @uklotzde's review comments above?

I would prefer to not add yet another function to the API for this one niche use case.

File: src/controllers/hid/hidcontroller.h [hunk 3]
```diff
@@ -75,7 +75,7 @@ class HidControllerJSProxy : public ControllerJSProxy {
     Q_INVOKABLE void send(const QList<int>& dataList,
             unsigned int length,
             const quint8& reportID,
-            bool resendUnchangedReport = false) {
+            const bool& resendUnchangedReport = false) {
```
Comment (Swiftb0y) [inline src/controllers/hid/hidcontroller.h:78]: passing a cheap to copy value type by reference does not make much sense.

File: src/controllers/hid/hidcontroller.h [hunk 4]
```diff
@@ -60,45 +57,79 @@ class HidControllerJSProxy : public ControllerJSProxy {
...
+        Q_UNUSED(length);
+        this->send(dataList, 0, 0);
+    }
+
+    /// @brief Sends HID OutputReport to HID device
+    /// @param dataList Data to send as list of bytes
+    /// @param length Unused but mandatory argument
+    /// @param reportID 1...255 for HID devices that uses ReportIDs - or 0 for devices, which don't use ReportIDs
+    /// @param resendUnchangedReport If set, the report will also be send, if the data are unchanged since last sending
+    Q_INVOKABLE void send(const QList<int>& dataList,
```
Comment (daschuer) [inline src/controllers/hid/hidcontroller.h:75]: You have changed the signature now from 

```
Q_INVOKABLE void send(const QList<int>& data, unsigned int length, unsigned int reportID) {
```

```
Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length, quint8 reportID, bool esendUnchangedReport = false)
```

But the function is deprecated anyway. Can this cause any compatibility issues? If in doubt I would prefer no to touch it.

File: src/controllers/hid/hidcontroller.h [hunk 5]
```diff
@@ -60,45 +57,79 @@ class HidControllerJSProxy : public ControllerJSProxy {
               m_pHidController(m_pController) {
     }
 
-    Q_INVOKABLE void send(const QList<int>& data, unsigned int length = 0) override {
-        m_pHidController->send(data, length);
+    /// @brief Sends HID OutputReport with hard coded ReportID 0 to HID device
+    ///        This function only works with HID devices, which don't use ReportIDs
+    /// @param dataList Data to send as list of bytes
+    /// @param length Unused optional argument
+    Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length = 0) override {
```
Comment (daschuer) [inline src/controllers/hid/hidcontroller.h:64]: How does this comment relate: 
https://github.com/mixxxdj/mixxx/blob/f2d670edeb483efb066bf7f00255d01df2f0017c/res/controllers/common-hid-packet-parser.js#L467

Introduced here: 
https://github.com/mixxxdj/mixxx/pull/1953

**Comment by daschuer** [comment]
Does it make sense to adjust res/controllers/common-hid-packet-parser.js to the new API?

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by JoergAtGithub** [comment]
> Does it make sense to adjust res/controllers/common-hid-packet-parser.js to the new API?

In fact, this is what I had done for testing this PR. There is only one single line to change: https://github.com/mixxxdj/mixxx/blob/2c3b119d354c05da1c3caf1c11cb0937e415481a/res/controllers/common-hid-packet-parser.js#L694
this needs to be changed to:
`controller.sendOutputReport(this.reportId, Uint8Array.from(data).buffer);`

File: src/controllers/hid/hidcontroller.h (hunk 5 above)
Comment (JoergAtGithub) [reply src/controllers/hid/hidcontroller.h:64]: These are different tings:
There are two categories of HID devices:

1. Devices that use Report-IDs in no report (than the device can have only one report of each type Input/Output/Feature).
   hidapi expect the Report-ID byte for Output and Feature-Reports to be 0 in this case, but don't prepend this 0 to Input-Reports itself.

2. Devices that use a Report-ID in any report (allowed range of report IDs is 1...255).

Therefore you can use the sendOutputReport/sendFeatureReport functions and specify the Report-ID argument as 0, if you want send data to an HID device, that don't uses Report-IDs.

File: src/controllers/hid/hidcontroller.h (hunk 4 above)
Comment (JoergAtGithub) [reply src/controllers/hid/hidcontroller.h:75]: I tested it with the unmodified common-hid-packet-parser.js and it works flawless, with and without the optional argument.

Comment (daschuer) [reply src/controllers/hid/hidcontroller.h:75]: Thank you for the double check!

File: src/controllers/hid/hidcontroller.h (hunk 5 above)
Comment (daschuer) [reply src/controllers/hid/hidcontroller.h:64]: Ah OK, that makes sense.

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
> In fact, this is what I had done for testing this PR. There is only one single line to change: 

For my understanding this saves some CPU cycles due to the QByteArray, right? Would you mind to add this commit? 

This PR looks good to me. 

@uklotzde: Can we merge?

**Comment by JoergAtGithub** [comment]
> For my understanding this saves some CPU cycles due to the QByteArray, right? Would you mind to add this commit?

I don't know, what's the more efficient implementation. I would prefer to do it in a follow-up PR for common-hid-packet-parser.js

**Review by uklotzde — APPROVED** [review]
Thank you for this follow-up PR. Reordering the parameters to reflect the order within the actual data make sense. Not touching the Q_INVOKABLE functions goes without saying.

The data is copied twice into QByteArray? I didn't check if this behavior is unchanged from before.

LGTM

**Pull request merged**